---
title: "Pathway activity inference from scRNA-seq"
author:
  - name: Pau Badia-i-Mompel
    affiliation:
    - Heidelberg Universiy
output: 
  BiocStyle::html_document:
    self_contained: true
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: show
package: "`r pkg_ver('decoupleR')`"
vignette: >
  %\VignetteIndexEntry{Pathway activity activity inference from scRNA-seq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

scRNA-seq yield many molecular readouts that are hard to interpret by
themselves. One way of summarizing this information is by inferring pathway
activities from prior knowledge.

In this notebook we showcase how to use `decoupleR` for pathway activity inference
with the 3k PBMCs 10X data-set. The data consists of 3k PBMCs from a Healthy
Donor and is freely available from 10x Genomics 
[here](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz)
from this [webpage](https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k).

# Loading packages
First, we need to load the relevant packages, `Seurat` to handle scRNA-seq data
and `decoupleR` to use statistical methods.

```{r "load packages", message = FALSE}
## We load the required packages
library(Seurat)
library(decoupleR)
library(SeuratData)

# Only needed for data handling and plotting
library(tidyverse)
library(patchwork)
library(ggplot2)
library(pheatmap)
```

# Single-cell processing

## Loading the data-set
We can download and open the data using `SeuratData`:
```{r "load data", eval=FALSE}
# Download
InstallData("pbmc3k")

# Load
data("pbmc3k")
pbmc3k
```

## QC
Here we follow the standard pre-processing steps as described in the `Seurat` [vignette](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). These
steps carry out the selection and filtration of cells based on quality control
metrics, the data normalization, the detection of highly variable features and
data scaling.

```{r "preprocessing", message=FALSE}
## Identification of mithocondrial genes
pbmc3k[["percent.mt"]] <- PercentageFeatureSet(pbmc3k, pattern = "^MT-")

## Filtering cells following standard QC criteria.
pbmc3k <- subset(pbmc3k, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & 
    percent.mt < 5)

## Normalizing the data
pbmc3k <- NormalizeData(pbmc3k, normalization.method = "LogNormalize", 
    scale.factor = 10000)
pbmc3k <- NormalizeData(pbmc3k)

## Identify the 2000 most highly variable genes
pbmc3k <- FindVariableFeatures(pbmc3k, selection.method = "vst", nfeatures = 2000)

## We scale the data prior to clustering
all.genes <- rownames(pbmc3k)
pbmc3k <- ScaleData(pbmc3k, features = all.genes)
```

## Projection, clustering and annotation
Then we group cells based on the similarity of their transcription profiles and
annotate the obtained clusters based on their marker genes.
To visualize the communities we perform UMAP reduction.

```{r "clustering", message=FALSE, warning=FALSE}
# Dimensional reduction
pbmc3k <- RunPCA(pbmc3k, features = VariableFeatures(object = pbmc3k), 
               verbose = FALSE)

# Find Nearest Neighbors
pbmc3k <- FindNeighbors(pbmc3k, dims = 1:10, verbose = FALSE)

# Cluster cells
pbmc3k <- FindClusters(pbmc3k, resolution = 0.5, verbose = FALSE)

# Project into UMAP
pbmc3k <- RunUMAP(pbmc3k, dims = 1:10, umap.method = "uwot", metric = "cosine")

## Assigning cell type identity to clusters (following Seurat's vignette)
new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD14+ Mono", "B", "CD8 T",
                     "FCGR3A+ Mono", "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc3k)
pbmc3k <- RenameIdents(pbmc3k, new.cluster.ids)
```

```{r "umap", message = FALSE, warning = FALSE}
DimPlot(pbmc3k, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

# PROGENy model
[PROGENy](https://saezlab.github.io/progeny/) is a comprehensive resource
containing a curated collection of pathways and their target genes, with weights
for each interaction. For this example we will use the human weights
(mouse is also available) and we will use the top 100 responsive genes ranked
by p-value. To access it we can use `decoupleR`.

```{r "progeny", message=FALSE}
net <- get_progeny(organism = 'human', top = 100)
net
```

# Activity inference with Weighted Mean
To infer activities we will run the Weighted Mean method (`wmean`). It infers
regulator activities by first multiplying each target feature by its associated
weight which then are summed to an enrichment score `wmean`. Furthermore,
permutations of random target features can be performed to obtain a null
distribution that can be used to compute a z-score `norm_wmean`, or a corrected
estimate `corr_wmean` by multiplying `wmean` by the minus log10 of the obtained
empirical p-value.

In this example we use `wmean` but we could have used any other.
To see what methods are available use `show_methods()`.

To run `decoupleR` methods, we need an input matrix (`mat`), an input prior
knowledge network/resource (`net`), and the name of the columns of net that we
want to use.

```{r "wmean", message=FALSE}
# Extract the normalized log-transformed counts
mat <- as.matrix(pbmc3k@assays$RNA@data)

# Run wmean
acts <- run_wmean(mat=mat, net=net, .source='source', .target='target',
                  .mor='weight', times = 100, minsize = 5)
acts
```

# Visualization
From the obtained results, we will select the `norm_wmean` activities and store
them in our object as a new assay called `pathways_wmean`:
```{r "new_assay", message=FALSE}
# Extract norm_wmean and store it in pathways_wmean in pbmc
pbmc3k[['pathways_wmean']] <- acts %>%
  filter(statistic == 'norm_wmean') %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)

# Change assay
DefaultAssay(object = pbmc3k) <- "pathways_wmean"

# Scale the data
pbmc3k <- ScaleData(pbmc3k)
pbmc3k@assays$pathways_wmean@data <- pbmc3k@assays$pathways_wmean@scale.data
```

This new assay can be used to plot activities. Here we visualize the Trail
pathway, associated with apoptosis, which seems that in B and NK cells is more
active.
```{r "projected_acts", message = FALSE, warning = FALSE, fig.width = 8, fig.height = 4}
p1 <- DimPlot(pbmc3k, reduction = "umap", label = TRUE, pt.size = 0.5) + 
  NoLegend() + ggtitle('Cell types')
p2 <- (FeaturePlot(pbmc3k, features = c("Trail")) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle('Trail activity')
p1 | p2
```

# Exploration
We can also see what is the mean activity per group across pathways:
```{r "mean_acts", message = FALSE, warning = FALSE}
# Extract activities from object as a long dataframe
df <- t(as.matrix(pbmc3k@assays$pathways_wmean@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Idents(pbmc3k)) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Transform to wide matrix
top_acts_mat <- df %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color pallette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks) 
```
In this specific example, we can observe that WNT to be more active in
Platelets, and that Trail is more active in B and NK cells.

## Session information

```{r session_info, echo=FALSE}
options(width = 120)
sessioninfo::session_info()
```